File: api/routes/jobs.py
"""POST /jobs + GET /jobs/{id} — in-process async job execution.

Replaces ui/main_window.py's _CallableWorker(QThread) pattern for the web
API: run_resume_ingestion()/run_cv_generation() are fully synchronous
(blocking LLM SDK calls), so asyncio's run_in_executor (default
ThreadPoolExecutor) is the async-native equivalent of _CallableWorker's
QThread — get blocking work off the thread handling the HTTP request,
same two-tier `except AppError / except Exception` catch shape as
_CallableWorker.run().

Job state lives in an in-memory dict guarded by a threading.Lock — no
Celery/Redis, per docs/technology_stack.md's single-user/local scope for
this phase. Lost on process restart: job records are ephemeral
orchestration metadata, not user data (Candidate/Evidence, which do need to
survive a restart, live in SQLite instead — see db/models.py).
"""

from __future__ import annotations

import asyncio
import tempfile
import threading
import uuid
from datetime import datetime, timezone
from pathlib import Path
from typing import Annotated, Literal, Union

from fastapi import APIRouter, Depends, HTTPException, UploadFile
from pydantic import BaseModel, Field

import app.errors as app_errors
from api.deps import get_services
from app.pipeline import (
    run_cv_generation,
    run_gap_recheck,
    run_resume_ingestion,
    run_skill_evidence_relink,
)
from app.resume_reader import read_resume_text
from app.services import Services
from domain.models import (
    AssembledCV,
    BulletProvenanceReport,
    Candidate,
    GenerationTiming,
    MatchResult,
    Requirement,
    Vacancy,
)

router = APIRouter(prefix="/jobs", tags=["jobs"])


class IngestJobRequest(BaseModel):
    type: Literal["ingest"] = "ingest"
    resume_text: str


class GenerateJobRequest(BaseModel):
    type: Literal["generate"] = "generate"
    candidate_id: str
    vacancy_text: str
    # A Regenerate/Regenerate as New on an existing CVDraft already has this
    # draft's own previously-parsed Vacancy on hand — passing it lets
    # run_cv_generation skip re-running the JD Parser stage entirely when
    # `vacancy_text` is unchanged from `existing_vacancy.raw_text` (the
    # common case: re-tailoring against the same job description, not a
    # new/edited one). `None` for a first-time generate (ExportScreen.tsx's
    # "Tailor CV to Job Description" — no prior parse exists yet).
    existing_vacancy: Vacancy | None = None


class RecheckJobRequest(BaseModel):
    type: Literal["recheck"] = "recheck"
    candidate_id: str
    requirements: list[Requirement]
    # Ids of Evidence items whose bullet the user has since toggled off in
    # the A4 editor.
    excluded_evidence_ids: list[str] = Field(default_factory=list)
    # The four fields below carry the document's own current content into
    # the recheck — see run_gap_recheck's docstring for the full reasoning
    # and frontend/src/lib/documentEvidence.ts /
    # frontend/src/lib/structuredDocument.ts for how the frontend builds
    # them from the live `DocumentModel` right before calling this route.
    edited_bullet_text: dict[str, str] = Field(default_factory=dict)
    manual_bullet_text: list[str] = Field(default_factory=list)
    document_skills: list[str] | None = None
    document_technologies: list[str] | None = None


class RelinkSkillEvidenceJobRequest(BaseModel):
    type: Literal["relink_skill_evidence"] = "relink_skill_evidence"
    candidate_id: str


JobRequest = Annotated[
    Union[IngestJobRequest, GenerateJobRequest, RecheckJobRequest, RelinkSkillEvidenceJobRequest],
    Field(discriminator="type"),
]


class IngestJobResult(BaseModel):
    candidate_id: str
    candidate: Candidate


class GenerateJobResult(BaseModel):
    vacancy: Vacancy
    match_result: MatchResult
    assembled_cv: AssembledCV
    provenance: BulletProvenanceReport
    timing: GenerationTiming


class RecheckJobResult(BaseModel):
    match_result: MatchResult


class RelinkSkillEvidenceJobResult(BaseModel):
    updated_skill_count: int
    updated_technology_count: int


class JobErrorBody(BaseModel):
    category: str
    message: str


class JobStatus(BaseModel):
    id: str
    type: Literal["ingest", "generate", "recheck", "relink_skill_evidence"]
    status: Literal["pending", "running", "succeeded", "failed"] = "pending"
    # Progress within a "generate" job's six-stage chain (app/pipeline.py's
    # GENERATION_STAGES) — `None` for every other job type, and for
    # "generate" itself until the first stage actually starts (there's a
    # real, if usually brief, gap between this record's creation and
    # _execute_generate picking it up from the executor). Frontend polling
    # is already 1s (useJobPolling), same cadence a stage takes at minimum,
    # so this is cheap to add and doesn't need its own faster poll.
    stage: str | None = None
    stage_number: int | None = None
    stage_count: int | None = None
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    finished_at: datetime | None = None
    result: IngestJobResult | GenerateJobResult | RecheckJobResult | RelinkSkillEvidenceJobResult | None = None
    error: JobErrorBody | None = None


_jobs: dict[str, JobStatus] = {}
_jobs_lock = threading.Lock()


def _execute_ingest(job_id: str, request: IngestJobRequest, services: Services) -> None:
    with _jobs_lock:
        _jobs[job_id].status = "running"
    try:
        candidate_id, candidate = run_resume_ingestion(request.resume_text, services)
    except app_errors.AppError as exc:
        _set_failed(job_id, exc)
        return
    except Exception as exc:  # noqa: BLE001 — mirrors _CallableWorker's last-resort catch
        _set_failed(job_id, exc)
        return
    _set_succeeded(job_id, IngestJobResult(candidate_id=candidate_id, candidate=candidate))


def _report_stage(job_id: str, label: str, number: int, count: int) -> None:
    with _jobs_lock:
        record = _jobs.get(job_id)
        if record is not None:
            record.stage = label
            record.stage_number = number
            record.stage_count = count


def _execute_generate(job_id: str, request: GenerateJobRequest, services: Services) -> None:
    with _jobs_lock:
        _jobs[job_id].status = "running"
    try:
        service = services.candidate_registry.service_for(request.candidate_id)
        candidate = service.get()
        if candidate is None:
            raise app_errors.ValidationError(f"Candidate {request.candidate_id!r} not found.")
        evidence = service.get_evidence()
        result = run_cv_generation(
            candidate,
            evidence,
            request.vacancy_text,
            services,
            existing_vacancy=request.existing_vacancy,
            on_stage=lambda label, number, count: _report_stage(job_id, label, number, count),
        )
    except app_errors.AppError as exc:
        _set_failed(job_id, exc)
        return
    except Exception as exc:  # noqa: BLE001
        _set_failed(job_id, exc)
        return
    _set_succeeded(
        job_id,
        GenerateJobResult(
            vacancy=result.vacancy,
            match_result=result.match_result,
            assembled_cv=result.assembled_cv,
            provenance=result.provenance,
            timing=result.timing,
        ),
    )


def _execute_recheck(job_id: str, request: RecheckJobRequest, services: Services) -> None:
    with _jobs_lock:
        _jobs[job_id].status = "running"
    try:
        service = services.candidate_registry.service_for(request.candidate_id)
        candidate = service.get()
        if candidate is None:
            raise app_errors.ValidationError(f"Candidate {request.candidate_id!r} not found.")
        evidence = service.get_evidence()
        match_result = run_gap_recheck(
            candidate,
            evidence,
            request.requirements,
            set(request.excluded_evidence_ids),
            services,
            edited_bullet_text=request.edited_bullet_text,
            manual_bullet_text=request.manual_bullet_text,
            document_skills=request.document_skills,
            document_technologies=request.document_technologies,
        )
    except app_errors.AppError as exc:
        _set_failed(job_id, exc)
        return
    except Exception as exc:  # noqa: BLE001
        _set_failed(job_id, exc)
        return
    _set_succeeded(job_id, RecheckJobResult(match_result=match_result))


def _execute_relink_skill_evidence(
    job_id: str, request: RelinkSkillEvidenceJobRequest, services: Services
) -> None:
    with _jobs_lock:
        _jobs[job_id].status = "running"
    try:
        service = services.candidate_registry.service_for(request.candidate_id)
        candidate = service.get()
        if candidate is None:
            raise app_errors.ValidationError(f"Candidate {request.candidate_id!r} not found.")
        evidence = service.get_evidence()
        skill_count, technology_count = run_skill_evidence_relink(service, candidate, evidence, services)
    except app_errors.AppError as exc:
        _set_failed(job_id, exc)
        return
    except Exception as exc:  # noqa: BLE001
        _set_failed(job_id, exc)
        return
    _set_succeeded(
        job_id,
        RelinkSkillEvidenceJobResult(updated_skill_count=skill_count, updated_technology_count=technology_count),
    )


def _set_succeeded(
    job_id: str,
    result: IngestJobResult | GenerateJobResult | RecheckJobResult | RelinkSkillEvidenceJobResult,
) -> None:
    with _jobs_lock:
        record = _jobs[job_id]
        record.status = "succeeded"
        record.result = result
        record.finished_at = datetime.now(timezone.utc)


def _set_failed(job_id: str, exc: Exception) -> None:
    with _jobs_lock:
        record = _jobs[job_id]
        record.status = "failed"
        record.error = JobErrorBody(category=type(exc).__name__, message=str(exc))
        record.finished_at = datetime.now(timezone.utc)


@router.post("", status_code=202)
async def create_job(request: JobRequest, services: Services = Depends(get_services)) -> JobStatus:
    job_id = uuid.uuid4().hex[:8]
    with _jobs_lock:
        _jobs[job_id] = JobStatus(id=job_id, type=request.type)

    loop = asyncio.get_event_loop()
    if isinstance(request, IngestJobRequest):
        loop.run_in_executor(None, _execute_ingest, job_id, request, services)
    elif isinstance(request, GenerateJobRequest):
        loop.run_in_executor(None, _execute_generate, job_id, request, services)
    elif isinstance(request, RecheckJobRequest):
        loop.run_in_executor(None, _execute_recheck, job_id, request, services)
    else:
        loop.run_in_executor(None, _execute_relink_skill_evidence, job_id, request, services)

    with _jobs_lock:
        return _jobs[job_id]


@router.post("/ingest-upload", status_code=202)
async def create_ingest_upload_job(
    file: UploadFile, services: Services = Depends(get_services)
) -> JobStatus:
    """Multipart-upload counterpart to `POST /jobs {"type": "ingest", ...}`
    (Version 4, Phase 4.5) — the frontend's paste-a-resume box gets a
    file-upload alternative, not a parallel ingestion pipeline. Extraction
    (app/resume_reader.py's PDF/DOCX/RTF/TXT/MD dispatch) runs against a
    temp copy of the upload, off the event loop via `run_in_executor` —
    same reasoning as every other blocking call in this file (see the
    module docstring), not just the LLM-calling ones: `pypdf.extract_text()`
    is normally milliseconds, but a pathological file (a huge embedded
    image, a deeply nested object graph) could take a while, and running it
    in-request on the event loop directly (as this used to) would block
    every other request the whole server is handling for that long, not
    just this one — worse than a slow upload, indistinguishable from the
    app itself going unresponsive. An unsupported type, unreadable file, or
    empty extraction raises ParsingError, which propagates out of
    run_in_executor's awaited future same as a synchronous raise would;
    api/errors.py's handler turns it into an immediate error response —
    that's a request-validation problem, not something that fails mid-
    LLM-call the way a genuine ingestion failure does, so it's surfaced
    before a job record even exists rather than as a "failed" job. Once
    extraction succeeds, this hands off to create_job() unchanged — same
    job id/polling/result shape a pasted-text ingest gets.
    """
    suffix = Path(file.filename or "").suffix
    with tempfile.NamedTemporaryFile(suffix=suffix, delete=False) as tmp:
        tmp.write(await file.read())
        tmp_path = Path(tmp.name)
    try:
        loop = asyncio.get_event_loop()
        resume_text = await loop.run_in_executor(None, read_resume_text, tmp_path)
    finally:
        tmp_path.unlink(missing_ok=True)
    return await create_job(IngestJobRequest(resume_text=resume_text), services)


@router.get("/{job_id}")
async def get_job(job_id: str) -> JobStatus:
    with _jobs_lock:
        record = _jobs.get(job_id)
    if record is None:
        raise HTTPException(status_code=404, detail=f"Job {job_id!r} not found.")
    return record

File: frontend/src/api/jobs.ts
// Ingest and generate are both POST /jobs { type, ... } -> poll GET
// /jobs/{id} until a terminal status — the async-native equivalent of
// ui/main_window.py's _CallableWorker(QThread) pattern. useJobPolling is
// the one genuinely nontrivial piece, shared by both purpose-built hooks
// below so IngestPanel/VacancyPanel+GenerateButton don't each have to
// discriminate the JobStatus.result union by hand.

import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { useEffect, useState } from "react";

import { ApiError, apiFetch, apiFetchUpload } from "./client";
import type {
  GenerateJobResult,
  IngestJobResult,
  JobErrorBody,
  JobStatus,
  RecheckJobResult,
  RelinkSkillEvidenceJobResult,
  Requirement,
  Vacancy,
} from "./models";

function isTerminal(status: JobStatus["status"] | undefined): boolean {
  return status === "succeeded" || status === "failed";
}

function useJobPolling(jobId: string | null) {
  return useQuery({
    queryKey: ["jobs", jobId],
    queryFn: () => apiFetch<JobStatus>(`/jobs/${jobId}`),
    enabled: jobId != null,
    // Poll every 1s while pending/running (matches the multi-second LLM
    // call latency api/routes/jobs.py's docstring describes); stop once a
    // terminal status is reached — TanStack Query v5's function-form
    // refetchInterval is exactly built for "poll until terminal state".
    refetchInterval: (query) => (isTerminal(query.state.data?.status) ? false : 1000),
  });
}

export function useIngestJob() {
  const [jobId, setJobId] = useState<string | null>(null);
  const queryClient = useQueryClient();

  const mutation = useMutation({
    mutationFn: (resumeText: string) =>
      apiFetch<JobStatus>("/jobs", {
        method: "POST",
        body: JSON.stringify({ type: "ingest", resume_text: resumeText }),
      }),
    onSuccess: (job) => setJobId(job.id),
  });

  const poll = useJobPolling(jobId);
  const status = poll.data?.status;

  useEffect(() => {
    if (status === "succeeded") {
      // Mirrors _on_ingest_succeeded's _reload_profiles(select_id=...) —
      // the picker's candidate list must include the just-created profile.
      void queryClient.invalidateQueries({ queryKey: ["candidates"] });
    }
  }, [status, queryClient]);

  return {
    ingest: (resumeText: string) => {
      setJobId(null);
      mutation.mutate(resumeText);
    },
    isSubmitting: mutation.isPending,
    isRunning: jobId != null && !isTerminal(status),
    result: status === "succeeded" ? (poll.data?.result as IngestJobResult) : undefined,
    error: status === "failed" ? (poll.data?.error as JobErrorBody | undefined) : undefined,
  };
}

// Version 4, Phase 4.5 — the file-upload counterpart to useIngestJob, for
// api/routes/jobs.py's POST /jobs/ingest-upload. Mirrors useIngestJob's
// shape exactly (same result/error types — both routes return the same
// IngestJobResult-shaped job); the only difference is the POST itself uses
// apiFetchUpload with a FormData body instead of apiFetch with a JSON one.
export function useIngestUploadJob() {
  const [jobId, setJobId] = useState<string | null>(null);
  const queryClient = useQueryClient();

  const mutation = useMutation({
    mutationFn: (file: File) => {
      const formData = new FormData();
      formData.append("file", file);
      return apiFetchUpload<JobStatus>("/jobs/ingest-upload", formData);
    },
    onSuccess: (job) => setJobId(job.id),
  });

  const poll = useJobPolling(jobId);
  const status = poll.data?.status;

  useEffect(() => {
    if (status === "succeeded") {
      void queryClient.invalidateQueries({ queryKey: ["candidates"] });
    }
  }, [status, queryClient]);

  // Unlike a pasted-text ingest, extraction (app/resume_reader.py) runs
  // synchronously in api/routes/jobs.py's POST /jobs/ingest-upload itself
  // — an unsupported type or unreadable file fails that POST directly
  // (mutation.error), no job record ever gets created to later transition
  // to "failed" the way poll.data?.error below assumes. Both are surfaced
  // through the same `error` shape so IngestPanel's caller doesn't need to
  // know which of the two failure paths actually happened.
  const uploadError =
    mutation.error instanceof ApiError
      ? { category: mutation.error.category, message: mutation.error.message }
      : undefined;

  return {
    ingest: (file: File) => {
      setJobId(null);
      mutation.mutate(file);
    },
    isSubmitting: mutation.isPending,
    isRunning: jobId != null && !isTerminal(status),
    result: status === "succeeded" ? (poll.data?.result as IngestJobResult) : undefined,
    error: status === "failed" ? (poll.data?.error as JobErrorBody | undefined) : uploadError,
  };
}

export function useGenerateJob() {
  const [jobId, setJobId] = useState<string | null>(null);

  const mutation = useMutation({
    mutationFn: (payload: { candidateId: string; vacancyText: string; existingVacancy?: Vacancy }) =>
      apiFetch<JobStatus>("/jobs", {
        method: "POST",
        body: JSON.stringify({
          type: "generate",
          candidate_id: payload.candidateId,
          vacancy_text: payload.vacancyText,
          existing_vacancy: payload.existingVacancy ?? null,
        }),
      }),
    onSuccess: (job) => setJobId(job.id),
  });

  const poll = useJobPolling(jobId);
  const status = poll.data?.status;

  return {
    // `existingVacancy` — a Regenerate/Regenerate as New on an existing
    // draft passes that draft's own already-parsed Vacancy, so the backend
    // can skip re-running the JD Parser stage when `vacancyText` matches
    // its `raw_text` unchanged (app/pipeline.py::run_cv_generation's own
    // `existing_vacancy` docstring). Omitted for a first-time generate
    // (ExportScreen.tsx) — there's no prior parse to reuse.
    generate: (candidateId: string, vacancyText: string, existingVacancy?: Vacancy) => {
      setJobId(null);
      mutation.mutate({ candidateId, vacancyText, existingVacancy });
    },
    isSubmitting: mutation.isPending,
    isRunning: jobId != null && !isTerminal(status),
    // Which of app/pipeline.py's six GENERATION_STAGES is currently
    // running — undefined before the first stage starts (a real, if
    // usually brief, gap right after the job record is created) and,
    // deliberately, still populated once the job finishes (the caller
    // only reads this while `isRunning`, and holding the last value
    // avoids a "Step 6/6" flash to blank right before `result` appears).
    progress:
      poll.data?.stage != null && poll.data.stage_number != null && poll.data.stage_count != null
        ? { stage: poll.data.stage, stageNumber: poll.data.stage_number, stageCount: poll.data.stage_count }
        : undefined,
    result: status === "succeeded" ? (poll.data?.result as GenerateJobResult) : undefined,
    error: status === "failed" ? (poll.data?.error as JobErrorBody | undefined) : undefined,
  };
}

// Phase 18 — re-runs just the Matching stage (app/pipeline.py::
// run_gap_recheck) against the CV as currently edited, not the full
// ingest/generate chain. Mirrors useGenerateJob's shape exactly; the only
// difference is the request payload and result type.
// Post-4.10 fixes round 3 — the recheck payload's `RecheckDocumentContext`
// carries the document's own current content into the recheck (see
// api/routes/jobs.py's RecheckJobRequest and app/pipeline.py's
// run_gap_recheck docstring for the full reasoning): edited/hand-typed
// bullet text and the document's own Skills/Technologies lists, none of
// which the stored Candidate profile alone can reflect once the person
// starts editing the draft directly.
export interface RecheckDocumentContext {
  editedBulletText: Record<string, string>;
  manualBulletText: string[];
  documentSkills: string[];
  documentTechnologies: string[];
}

export function useRecheckGapsJob() {
  const [jobId, setJobId] = useState<string | null>(null);

  const mutation = useMutation({
    mutationFn: (payload: {
      candidateId: string;
      requirements: Requirement[];
      excludedEvidenceIds: string[];
      documentContext: RecheckDocumentContext;
    }) =>
      apiFetch<JobStatus>("/jobs", {
        method: "POST",
        body: JSON.stringify({
          type: "recheck",
          candidate_id: payload.candidateId,
          requirements: payload.requirements,
          excluded_evidence_ids: payload.excludedEvidenceIds,
          edited_bullet_text: payload.documentContext.editedBulletText,
          manual_bullet_text: payload.documentContext.manualBulletText,
          document_skills: payload.documentContext.documentSkills,
          document_technologies: payload.documentContext.documentTechnologies,
        }),
      }),
    onSuccess: (job) => setJobId(job.id),
  });

  const poll = useJobPolling(jobId);
  const status = poll.data?.status;

  return {
    recheck: (
      candidateId: string,
      requirements: Requirement[],
      excludedEvidenceIds: string[],
      documentContext: RecheckDocumentContext,
    ) => {
      setJobId(null);
      mutation.mutate({ candidateId, requirements, excludedEvidenceIds, documentContext });
    },
    isSubmitting: mutation.isPending,
    isRunning: jobId != null && !isTerminal(status),
    result: status === "succeeded" ? (poll.data?.result as RecheckJobResult) : undefined,
    error: status === "failed" ? (poll.data?.error as JobErrorBody | undefined) : undefined,
  };
}

// The on-demand "Re-evaluate Skill Dependencies" action
// (app/pipeline.py::run_skill_evidence_relink) — a standalone linking-only
// pass over the profile's current Skills/Technologies/Evidence, run
// whenever a person wants links refreshed rather than automatically on
// every edit. Mirrors useRecheckGapsJob's shape, plus an
// invalidate-on-success (mirroring useIngestJob's own) so the refreshed
// evidence_ids show up in the Skills/Technologies sections without a
// manual reload.
export function useRelinkSkillEvidenceJob() {
  const [jobId, setJobId] = useState<string | null>(null);
  const [candidateId, setCandidateId] = useState<string | null>(null);
  const queryClient = useQueryClient();

  const mutation = useMutation({
    mutationFn: (targetCandidateId: string) =>
      apiFetch<JobStatus>("/jobs", {
        method: "POST",
        body: JSON.stringify({ type: "relink_skill_evidence", candidate_id: targetCandidateId }),
      }),
    onSuccess: (job) => setJobId(job.id),
  });

  const poll = useJobPolling(jobId);
  const status = poll.data?.status;

  useEffect(() => {
    if (status === "succeeded" && candidateId != null) {
      void queryClient.invalidateQueries({ queryKey: ["candidates", candidateId] });
    }
  }, [status, candidateId, queryClient]);

  return {
    relink: (targetCandidateId: string) => {
      setJobId(null);
      setCandidateId(targetCandidateId);
      mutation.mutate(targetCandidateId);
    },
    isSubmitting: mutation.isPending,
    isRunning: jobId != null && !isTerminal(status),
    result: status === "succeeded" ? (poll.data?.result as RelinkSkillEvidenceJobResult) : undefined,
    error: status === "failed" ? (poll.data?.error as JobErrorBody | undefined) : undefined,
  };
}
